### PR TITLE
Use Shutdown to gracefully halt HTTP servers

### LINKED ISSUE #2902: Gracefully drain checkin requests on shutdown
We need to be able to rollout frequent releases of Fleet Server without causing undue disruptions visible to Elastic Agents that may be in a long polling connection to Fleet Server. Ideally, when the Fleet Server processes receives a SIGINT, all active checkin requests will get a normal 200 response as they normally do when the long poll expires.

We currently handle shutdowns via:
1. We have a signal handler installed here which will trigger the context to be cancelled: https://github.com/elastic/fleet-server/blob/0826ebc1f10562b2897859f2de3683e02b80270f/internal/pkg/signal/signal.go#L18
2. That context is passed to the `server.Run` function here: https://github.com/elastic/fleet-server/blob/0826ebc1f10562b2897859f2de3683e02b80270f/cmd/fleet/main.go#L125
3. `http.Server.Close()` is called when the server context is closed, here: https://github.com/elastic/fleet-server/blob/0826ebc1f10562b2897859f2de3683e02b80270f/internal/pkg/api/server.go#L85
4. The checkin handler simply returns the context cancellation reason without writing a response here: https://github.com/elastic/fleet-server/blob/0826ebc1f10562b2897859f2de3683e02b80270f/internal/pkg/api/handleCheckin.go#L242

From my reading of the Go http docs and the code, we need to:
- [ ] Add tests that verify this behavior at the API level (likely useful to start here for testing the implementation during dev TDD-style)
- [ ] Change (4) to log and break the loop instead of returning `ctx.Err()`
	- `request.Context()` that is used in (4) is using the context from (2) since it's passed in as the `BaseContext`. Therefore I think this will be cancelled before the connection is closed, which should allow us to write a response. 
- [ ] Use `http.Server.Shutdown()` (with a context that has a timeout) instead of `http.Server.Close()` at (2) in order to gracefully close connections once they're idle, rather than close them immediately.

Let's not worry about the other endpoints at this time.

### PR DESCRIPTION
**Pull request by michel-laterman**
## What is the problem this PR solves?

Checkin API long-poll connections terminate with an error when the server needs to shut down (providing a 5xx to the client).

## How does this PR solve the problem?

Use the Shutdown method with a timeout to try to gracefully shutdown
HTTP connections. If the server times out while attempting to drain the
connections, connections are forced closed. If there is a long-polling
checkin request, the server should send a 200 response with the ack
token used in the request (same behaviour as the poll completing
normally).

## Design Checklist

- [x] I have ensured my design is stateless and will work when multiple fleet-server instances are behind a load balancer.
- [ ] I have or intend to scale test my changes, ensuring it will work reliably with 100K+ agents connected.
- ~~I have included fail safe mechanisms to limit the load on fleet-server: rate limiting, circuit breakers, caching, load shedding, etc.~~

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/fleet-server#changelog)

## Related issues

- Closes #2902

### CODE REVIEW

**Comment by michel-laterman** [comment]
Does anyone know why the linter is all of a sudden having issues with our apm imports?

**Comment by jlind23** [comment]
@michel-laterman Looks like the linter is ok and only Sonar analysis complains about the insufficient code coverage. 
Here is the coverage report - Profile and handlecheckin doensn't have any test. I let you decide whether or not this is a real problem.
Screenshot 2023-12-15 at 07 46 04

**Comment by michel-laterman** [comment]
The changes to handleCheckin should be covered by the new integration test (but coverage for those tests don't work as expected). I've also added a test case for the profiler

**Review by jsoriano — APPROVED** [review]
Code-wise LGTM, not tested.

File: internal/pkg/api/server.go [hunk 1]
```diff
@@ -130,23 +113,29 @@ func (s *server) Run(ctx context.Context) error {
...
 
 	select {
+	// Listen and return any errors that occur from the server listener
 	case err := <-errCh:
 		if !errors.Is(err, context.Canceled) {
-			return err
+			return fmt.Errorf("error while serving API listener: %w", err)
+		}
+	// Do a clean shutdown if the context is cancelled
+	case <-ctx.Done():
```
[the remark this answers is not in the document]
Comment (michel-laterman) [reply internal/pkg/api/server.go:132]: We set the context to be the base context for requests, not connections. Some of our endpoints will use the (request) context for lifecycle control (mainly checkin).
Most other endpoints  are much quicker and use the bulker (which uses a different base context) and will not return early.
The shutdown called here will stop the server from accepting new connections, and give it some time to write all the responses

File: internal/pkg/api/handleCheckin.go [hunk 2]
```diff
@@ -319,7 +319,16 @@ func (ct *CheckinT) ProcessRequest(zlog zerolog.Logger, w http.ResponseWriter, r
 		for {
 			select {
 			case <-ctx.Done():
-				span.End()
+				defer span.End()
+				// If the request context is canceled, the API server is shutting down.
+				// We want to immediately stop the long-poll and return a 200 with the ackToken and no actions.
```
[the remark this answers is not in the document]
Comment (michel-laterman) [reply internal/pkg/api/handleCheckin.go:324]: yep, they will wait a little (iirc they have some jitter built in) then reconnect

**Comment by jlind23** [comment]
buildkite test it

**Comment by jlind23** [comment]
@nchaulet @michel-laterman all tests passed, shall we merge this?

**Review by nchaulet — APPROVED** [review]
Tested locally and it seems to work as expected 🚀

**Pull request merged**